File: Sources/JVoice/Services/Transcription/VocabularyPrompt.swift
import Foundation

/// Builds the decoder-conditioning prompt from the user's custom words.
///
/// Whisper conditions its decoder on "previous transcript" text. Feeding the
/// custom vocabulary as that text (OpenAI's `initial_prompt` technique) makes
/// the model strongly prefer those spellings when the audio sounds like them —
/// fixing recognition at the source instead of patching it afterwards.
public enum VocabularyPrompt {
    /// Cap on words included — keeps the decoder prefill cheap; prompt tokens
    /// linearly increase per-window decode cost.
    public static let maxWords = 40
    /// Hard cap on encoded tokens, well under WhisperKit's own
    /// `maxTokenContext/2 - 1` (~111) trim.
    public static let maxPromptTokens = 96

    /// The conditioning text, or nil when there is nothing to bias toward.
    public static func text(for words: [String]) -> String? {
        let cleaned = words
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        guard !cleaned.isEmpty else { return nil }
        // The leading space matters: Whisper's BPE merges a leading space into
        // word tokens, so conditioning text must look like natural transcript.
        return " " + cleaned.prefix(maxWords).joined(separator: ", ")
    }
}

File: windows/JVoice.Core/Text/VocabularyPrompt.cs
namespace JVoice.Core.Text;

/// Builds the decoder-conditioning prompt from the user's custom words
/// (OpenAI's `initial_prompt` technique). Ports VocabularyPrompt.swift verbatim.
public static class VocabularyPrompt
{
    /// Cap on words included — keeps the decoder prefill cheap.
    public const int MaxWords = 40;
    /// Hard cap on encoded tokens (consumed by the engine when tokenizing).
    public const int MaxPromptTokens = 96;

    /// The conditioning text, or null when there is nothing to bias toward.
    public static string? Text(IReadOnlyList<string> words)
    {
        var cleaned = new List<string>(words.Count);
        foreach (var w in words)
        {
            var t = w.Trim();
            if (t.Length > 0) cleaned.Add(t);
        }
        if (cleaned.Count == 0) return null;
        // Leading space matters: Whisper's BPE merges a leading space into word tokens.
        return " " + string.Join(", ", cleaned.Take(MaxWords));
    }
}
